src/callbacks_stage/positivity_zhang_shu_dg2d.jl

```diff
@@ -42,4 +42,48 @@ function limiter_zhang_shu!(u, threshold::Real, variable,
 
     return nothing
 end
+
+function limiter_zhang_shu!(u, threshold::Real, variable,
+                            mesh::Union{StructuredMesh{2}, StructuredMeshView{2},
+                                        UnstructuredMesh2D, P4estMesh{2},
+                                        T8codeMesh{2}},
+                            equations, dg::DGSEM, cache)
+    @unpack weights = dg.basis
+
+    @threaded for element in eachelement(dg, cache)
+        # determine minimum value
+        value_min = typemax(eltype(u))
+        for j in eachnode(dg), i in eachnode(dg)
+            u_node = get_node_vars(u, equations, dg, i, j, element)
+            value_min = min(value_min, variable(u_node, equations))
+        end
+
+        # detect if limiting is necessary
+        value_min < threshold || continue
+
+        # compute mean value
+        u_mean = zero(get_node_vars(u, equations, dg, 1, 1, element))
+        total_volume = zero(real(mesh))
+        for j in eachnode(dg), i in eachnode(dg)
+            jacobian_node = abs(inv(cache.elements.inverse_jacobian[i, j, element]))
+            u_node = get_node_vars(u, equations, dg, i, j, element)
+            u_mean += u_node * weights[i] * weights[j] * jacobian_node
+            total_volume += weights[i] * weights[j] * jacobian_node
+        end
+        # normalize with the total volume
+        u_mean = u_mean / total_volume
```

Comment:
It seems to me like this is the only part changed here.

If this is true, why not introduce an auxiliary function here that just computes the mean value, e.g.,
```julia
function mean_value_element(u, equations, dg, element, mesh::DispatchOnMe!)
   ...
```
and dispatch on the mesh type? Just a thought though...

Reply:
That is true, only those few lines are different. I like the idea of introducing an auxiliary function to avoid duplication. I will do this along with updating any test values that may change.

Reply:
Thanks!

Reply:
Is it even necessary to dispatch on different mesh types?
Since `cache.elements.inverse_jacobian` is even provided for `TreeMesh` we might still use the new implementation with `AbstractMesh`.

Reply:
The `TreeMesh` saves the constant value of the Jacobian at every node. So, that I just use the pointwise values available.

For instance when I run a Sedov blast on `TreeMesh` and check the `inverse_jacobian` value I get

```julia
julia> trixi_include("examples/tree_2d_dgsem/elixir_euler_sedov_blast_wave.jl", tspan=(0.0, 0.2))
julia> semi.cache.elements.inverse_jacobian
568-element Vector{Float64}:
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 ⋮
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
 8.0
```

Reply:
I'm just testing something like this
```
@inline function get_inverse_jacobian(inverse_jacobian, mesh::TreeMesh{2}, i, j, element)
    return inverse_jacobian[element]
end
@inline function get_inverse_jacobian(inverse_jacobian, mesh::Union{StructuredMesh{2}, ...}, i, j, element)
    return inverse_jacobian[i, j, element]
end
```

Reply:
> I'm just testing something like this
> 
> ```
> @inline function get_inverse_jacobian(inverse_jacobian, mesh::TreeMesh{2}, i, j, element)
>     return inverse_jacobian[element]
> end
> @inline function get_inverse_jacobian(inverse_jacobian, mesh::Union{StructuredMesh{2}, ...}, i, j, element)
>     return inverse_jacobian[i, j, element]
> end
> ```

Ah, now I understand what these helper functions would do. So where does these functions "live" in your implementation. Are they somewhere higher up the chain because it is agnostic to the limiter? E.g., in `src/solvers/dg.jl`?

Reply:
I'd say these should live in the top-level file for each solver type and dimension, e.g., `src/solvers/dgsem_tree/dg_2d.jl` for `TreeMesh` and `src/solvers/dgsem_structured/dg_2d.jl` for all curved meshes.

Reply:
I have not yet pushed them. Just tested them to get rid of the doubled code dispatching produces.

Reply:
If we want to implement them, we probably could implement a dimension-independent version like
```
@inline function get_inverse_jacobian(inverse_jacobian, mesh::TreeMesh, element, indices...)
    return inverse_jacobian[element]
end
@inline function get_inverse_jacobian(inverse_jacobian, mesh::Union{StructuredMesh, ...}, element, indices...)
    return inverse_jacobian[indices..., element]
end
```
Or include `element` to `indices`.

Reply:
Okay, I can prototype this and see how it looks in this PR.